File: templates/common/utils/pdf-extract/scripts/extract.ts
// jinhak-harness — common/utils/pdf-extract
// Node ↔ Python 결합: extract.py를 spawn해 JSON stdout을 파싱한다.
//
// 사용:
//   import { extract } from "./extract.ts";
//   const data = await extract("/path/to/file.pdf");
//
// 실행 시점에 Node 22+ --experimental-strip-types로 타입을 벗긴다.

import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

export interface PdfTable extends Array<Array<string | null>> {}
export interface PdfPage {
  text: string;
  tables: PdfTable[];
}
export interface PdfExtractResult {
  pages: PdfPage[];
}

const HERE = dirname(fileURLToPath(import.meta.url));
const PY_SCRIPT = resolve(HERE, "extract.py");

export function extract(
  pdfPath: string,
  options: { python?: string } = {},
): Promise<PdfExtractResult> {
  const python = options.python ?? process.env.JINHAK_PYTHON ?? "python3";

  return new Promise((res, rej) => {
    const child = spawn(python, [PY_SCRIPT, pdfPath], {
      stdio: ["ignore", "pipe", "pipe"],
    });

    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (chunk) => {
      stdout += chunk.toString("utf-8");
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk.toString("utf-8");
    });

    child.on("error", rej);
    child.on("close", (code) => {
      if (code !== 0) {
        rej(
          new Error(
            `extract.py exited with code ${code}: ${stderr.trim() || "(no stderr)"}`,
          ),
        );
        return;
      }
      try {
        res(JSON.parse(stdout) as PdfExtractResult);
      } catch (err) {
        rej(new Error(`failed to parse extract.py stdout as JSON: ${(err as Error).message}`));
      }
    });
  });
}

File: templates/common/utils/pdf-extract/scripts/extract.py
#!/usr/bin/env python3
"""
jinhak-harness — common/utils/pdf-extract
PDF 파일에서 페이지별 텍스트와 표를 뽑아 JSON으로 stdout에 출력한다.

사용:
    python3 extract.py /path/to/file.pdf

출력 스키마:
    {"pages": [{"text": "...", "tables": [[[..]]]}, ...]}
"""

from __future__ import annotations

import json
import sys
from pathlib import Path


def extract(pdf_path: Path) -> dict:
    import pdfplumber  # lazy import so --help 등 부수 명령에서 빠르게 종료

    pages: list[dict] = []
    with pdfplumber.open(str(pdf_path)) as pdf:
        for page in pdf.pages:
            text = page.extract_text() or ""
            try:
                tables = page.extract_tables() or []
            except Exception:
                tables = []
            pages.append({"text": text, "tables": tables})
    return {"pages": pages}


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print("usage: extract.py <pdf-path>", file=sys.stderr)
        return 2

    pdf_path = Path(argv[1])
    if not pdf_path.exists():
        print(f"file not found: {pdf_path}", file=sys.stderr)
        return 1

    result = extract(pdf_path)
    json.dump(result, sys.stdout, ensure_ascii=False)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
